dataset/src/main/kotlin/com/cosmotech/dataset/service/DatasetServiceImpl.kt

```diff
@@ -129,13 +333,221 @@ internal class DatasetServiceImpl(
       hasChanged = true
     }
 
+    dataset.main?.let {
+      existingDataset.main = it
+      hasChanged = true
+    }
+
     return if (hasChanged) {
       datasetRepository.save(existingDataset)
     } else {
       existingDataset
     }
   }
 
+  override fun twingraphQuery(
+      organizationId: String,
+      datasetId: String,
+      datasetTwinGraphQuery: DatasetTwinGraphQuery
+  ): String {
+    val dataset = getPrerequisiteDataset(organizationId, datasetId)
+    val resultSet =
+        trx(dataset) {
+          csmRedisGraph.query(
+              it.twingraphId!!,
+              datasetTwinGraphQuery.query,
+              csmPlatformProperties.twincache.queryTimeout)
+        }
+    return resultSet.toJsonString()
+  }
+
+  fun <T> trx(dataset: Dataset, actionLambda: (Dataset) -> T): T {
+    datasetRepository.save(dataset.apply { status = Dataset.Status.PENDING })
+    val result = actionLambda(dataset)
+    datasetRepository.save(dataset.apply { status = Dataset.Status.COMPLETED })
+    return result
+  }
+
+  override fun twingraphBatchUpdate(
+      organizationId: String,
+      datasetId: String,
+      twinGraphQuery: DatasetTwinGraphQuery,
+      body: Resource
+  ): TwinGraphBatchResult {
+
+    val dataset = getPrerequisiteDataset(organizationId, datasetId)
+    resourceScanner.scanMimeTypes(body, listOf("text/csv", "text/plain"))
+
+    val result = TwinGraphBatchResult(0, 0, mutableListOf())
+    trx(dataset) { localDataset ->
+      processCSVBatch(body.inputStream, twinGraphQuery, result) {
+        try {
+          csmRedisGraph.query(localDataset.twingraphId, it)
+          result.processedLines++
+        } catch (e: JedisDataException) {
+          result.errors.add("#${result.totalLines}: ${e.message}")
+        }
+      }
+    }
+    return result
+  }
+
+  @Suppress("ThrowsCount")
+  fun getPrerequisiteDataset(
+      organizationId: String,
+      datasetId: String,
+      status: Dataset.Status? = null,
+      query: String? = null
```

Comment:
Yes, it's a migration from twingraph to prevent update Query. Actually, we need to update but maybe sometimes not? Need to check & to discuss if we keep it